tests/D2L.CodeStyle.Analyzers.Test/Pinning/PinnedTestsBase.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System.Collections.Immutable;
+using D2L.CodeStyle.Analyzers.Pinning;
+using D2L.CodeStyle.Annotations.Pinning;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+using NUnit.Framework;
+
+namespace D2L.CodeStyle.Analyzers.Tests.Pinning {
+	internal abstract class PinnedTestsBase {
+
+		private ImmutableArray<MetadataReference> m_metadataReferences;
+		private ImmutableArray<DiagnosticAnalyzer> m_analyzers;
+		private CompilationWithAnalyzersOptions m_compilationWithAnalyzersOptions;
+
+		private const string ATTRIBUTE_DEFINITION = @"
+namespace System.Runtime.CompilerServices
+{
+    public static class IsExternalInit {}
+}
+";
+
+		private class StringAdditionalText : AdditionalText {
+			private readonly string m_text;
+
+			public StringAdditionalText( string path, string text ) {
+				Path = path;
+				m_text = text;
+			}
+
+			public override string Path { get; }
+
+			public override SourceText GetText(CancellationToken token) {
+				return SourceText.From( m_text );
+			}
+		}
+
+		public void SetUp( params DiagnosticAnalyzer[] analyzers ) {
+			List<MetadataReference> references = new List<MetadataReference> {
+				MetadataReference.CreateFromFile(typeof(CSharpCompilation).Assembly.Location),
+				MetadataReference.CreateFromFile(typeof(object).Assembly.Location),
+				MetadataReference.CreateFromFile(typeof(PinnedAttribute).Assembly.Location),
+				MetadataReference.CreateFromFile(typeof(PinnedAttributeAnalyzer).Assembly.Location, MetadataReferenceProperties.Assembly),
+			};
+			m_metadataReferences = references.ToImmutableArray();
+
+			m_analyzers = analyzers.ToImmutableArray();
+
+			AdditionalText allowedList = new StringAdditionalText( "UnpinnedAllowedList.txt", $@"System.Collections.Generic.Dictionary`2" );

```

Comment:
The other analyzers use it. I removed the other methods that aren't used, but left minor changes like this so that it was less work to do it as 3 PRs without missing anything, since some of the changes came from working on the must be pinned analyzer, but were for the recursive pinning.